Source Maps v4 Proposal (#538)
MicrosoftEdge/MSEdgeExplainers: DevTools/SourceMaps/explainer.md

Suppose I receive the following information from my application:

```
TypeError: Cannot read properties of undefined (reading 'x')
    at re.draw (bundle.js:1:16372)
    at re.drawLayer (bundle.js:1:14170)
    at be.draw (bundle.js:1:74592)
    at Te.render (bundle.js:1:114230)
    at Te.reflowCanvas (bundle.js:1:113897)
    at HTMLDocument.anonymous (bundle.js:1:135849)
```

The source map will resolve the first stack frame to `draw(src, sprite, x, y, frame)`, and will correctly point out that the failure here is that the undefined value is actually `frame`. However, *that is not the name of the function*. The function's name is `draw`, which is a member of the `Render` class.

If I simply apply the value of the `names` array to the function, decoding the stack trace is not particularly useful. It would look something like this:

```
TypeError: Cannot read properties of undefined (reading 'x')
    at frame (file1.ts:302:7)
    at draw (file1.ts:144:5)
    at Render (file2.ts:95:5)
    at drawArray (file3.ts:178:5)
    at render (file3.ts:155:5)
    at game (file4.ts:39:5)
```

> Using the source map to navigate the source code by hand, I can reconstruct the original call stack:

Reviewer: ```suggestion
The source line information here is improved and allows me to more easily find the source files and lines, but the function names are modified and now incorrect. Using the source map to navigate the source code by hand, I can reconstruct the original call stack including the class full names and anonymous functions:

```

I had to jump back and forth between these three stacks a few times to fully grock what was shifting (both good and bad) between them, so trying to add a touch more detail explaining the differences.

* **Anonymous callback name inference**: Callback functions are often anonymous, but the name of the function being invoked is typically interesting. In the following example, an interesting function name might be something along the lines of `anonymous function passed to myArray.map`:

```ts
    return myArray.map((item, index) => {
        /// ...
    });
```

 * **Computed names expressions**: Computed names also present an interesting problem of data that are not available at build time. As best as possible, the original source text should be simply preserved:

 ```ts
class Example {
    static [Symbol.species]() { return Object; }
    ['to' + 'String']() {
        return 'Hello, world.';
    }
}
 ```

* In the above example, we would want to see `static Example.[Symbol.species]` and `Example.['to' + 'String']` in the `scopeNames` field of the source map.

> 

Reviewer: ```suggestion
## Related Discussions
This proposal has a handful of discussions across related GitHub repos:
- [TypeScript Issue](https://github.com/microsoft/TypeScript/issues/46695)
- [TypeScript Prototype PR](https://github.com/microsoft/TypeScript/pull/47152)
- [WebPack Proposal](https://github.com/webpack/webpack/issues/14996)
```

There's a few related discussions that are in-depth enough to ensure readers of this explainer are aware of.

```

The source map will resolve the first stack frame to `draw(src, sprite, x, y, frame)`, and will correctly point out that the failure here is that the undefined value is actually `frame`. However, *that is not the name of the function*. The function's name is `draw`, which is a member of the `Render` class.

If I simply apply the value of the `names` array to the function, decoding the stack trace is not particularly useful. It would look something like this:

```
TypeError: Cannot read properties of undefined (reading 'x')
    at frame (file1.ts:302:7)
    at draw (file1.ts:144:5)
    at Render (file2.ts:95:5)
    at drawArray (file3.ts:178:5)
    at render (file3.ts:155:5)
    at game (file4.ts:39:5)
```

Using the source map to navigate the source code by hand, I can reconstruct the original call stack:

```
TypeError: Cannot read properties of undefined (reading 'x')
    at Render#draw (file1.ts:302:7)
    at Render#drawLayer (file1.ts:144:5)
    at GameObject#draw (file2.ts:95:5)
    at Game#render (file3.ts:178:5)
    at Game#reflowCanvas (file3.ts:155:5)
    at [anonymous function passed to addEventListener] (file4.ts:39:5)
```

> But doing this required that I dump the mappings and the source files from the source map and manually inspect the source files. In general, this requires that I use a library to parse the `mappings` field from Source Maps (because [as of Source Maps v3](https://sourcemaps.info/spec.html), this field is stored in a stateful way).

Author: It's not just that it's VLQ-encoded, as VLQs can be stateless. It's that many of the mappings in V3 are relative to the previous line entry. The comma- and semicolon-separation also are required. The new proposed `scopes` VLQs are a bit less stateful, but are still stateful from VLQ to VLQ entry.

> But doing this required that I dump the mappings and the source files from the source map and manually inspect the source files. In general, this requires that I use a library to parse the `mappings` field from Source Maps (because [as of Source Maps v3](https://sourcemaps.info/spec.html), this field is stored in a stateful way).

Reviewer 2: ```suggestion
But doing this required that I dump the mappings and the source files from the source map and manually inspect the source files. In general, this requires that I use a library to parse the `mappings` field from Source Maps (because [as of Source Maps v3](https://sourcemaps.info/spec.html), the position offsets of each mapping are relative to the one before).
```

## Proposed Solution

We want to adopt the general approach taken by the `pasta-sourcemaps` ("Pretty (and) Accurate Stack Trace Analysis") library. However, instead of adding to the `names` field, we will add two fields.

* Add two additional field to the source map: `scopes` and `scopeNames`
* The `scopes` field should be a list of ranges and pointers into `scopeNames`

> * The ranges in the field should always point to **Original Source** locations.

Reviewer 2: Untangling these from `names` is probably cleaner. Might be be worth speccing here the expectation of the sourcemap "apply" operation needed to combine sourcemaps? (considering that a sourcemap combine stage may be combining N earlier maps with 1 new map in the case of a transformation like inlining).

Author: Well, I was thinking about this, too. Since the original change, I pivoted that there's only one `scopes` field -- not an array of `scopes` fields -- but the bundler will need to process these either way. I'm working on that with webpack/webpack#14996.

```js
// example.min.js
1 (()=>{throw new Error('Intentional.')})();
```

Examining the call stack of this looks something like:

```
Error: Intentional.
    at <anonymous>:1:13
    at <anonymous>:1:40
```

We can get to the exact *locations* with source maps today, but assuming that the `mapping`'s 5th field ("the zero-based index into the `names` list associated with the segment) is encoded as it typically is, which is meant for live debugging, the decoded stack would actually look like this:

```
Error: Intentional
    at throw (example.js:2:5)
    at foo (example.js:4:1)
```

What we want to get to from here would be, at minimum:

```
Error: Intentional.
    at foo (example.js:2:5)
    at Global code (example.js:4:1)
```

### Improving the contents of the map

Let us re-examine the original code in this example:

```ts
1 function foo(bar: number): never {
2     throw new Error('Intentional.');
3 }
4 foo();
```

The scopes list should be as follows:

1. Line 1 Col 1 -> Line 1 Col 34 (captures `function foo(bar: number): never ` including the trailing space): `Global code`
1. Line 1 Col 35 -> Line 3 Col 1 (captures the entire body of `foo`): `foo`

> 1. Line 3 Col 2 -> Line 4 Col 7 (captures the newline + invocation of `foo`): `Global code`

Reviewer 2: Contrasting with pasta-sourcemaps:

Pasta uses "nested" scopes. Referring to ^ as "listed" scopes for this thread, equivalent would be:

1. Line 1 Col 1 -> Line 4 Col 7 (captures until end of invocation of `foo`): `<top-level>`
1. Line 1 Col 35 -> Line 3 Col 1 (captures the entire body of `foo`): `foo`

The [spec](https://github.com/bloomberg/pasta-sourcemaps/blob/9d2022adbd87c76af1d10a352963ff1765c52def/spec.md) doesn't seem to dwell on this, but some observations:

- "nested" is smaller
- Can't recursively bisect the "nested" `scopes` to find position
  - VLQ doesn't permit this anyway, must be linearly unpacked

Author: I'm not sure I understand what "nested" scopes are here, but regarding the second comment: the approach to bisecting the `scopes`:

 - Parse all of them
 - Sort by start line, then start column
 - Now you can binary search, you should look for the last entry that starts before and ends after the mapping position

Doesn't that meet your requirements?

Reviewer 2: Cross-post from https://github.com/webpack/webpack/issues/14996#issuecomment-996748709:

> ```
> // *nested*                            // *linear*
> | "<top-level>"                        | "<top-level>"
> |                                      |
> |  |  "causeError"         vs            |  "causeError"
> |  |                                     |
> |                                      | "<top-level>"
> ```

"Nested" scopes have one "entry" per scope in the source code.
"Linear" scopes have an entry each time you transition from one scope to another, left to right, top to bottom.

Nested will tend to result in fewer mappings, and a data structure that more semantically corresponds to the original source

Author: Ahhh, I understand! Actually, the way I implemented it in TypeScript was the "nested" version, and while I think I was meaning for this to indicate the "logical" view of the scopes list, but you're right -- I did put this into the expected list of these in the spec. I'll get that fixed.

By de-duplicating the names, that yields the expected `scopeNames` value of `['Global code', 'foo']`.

Finally, `scopes` will need to be an array of strings, and this array should have the same number of entries as the `sources` field. Each string will be a comma-delimited list of Base64-encoded VLQs with the following set of entries (please note, all offsets are 0-based, in concurrence with the previous Source Maps spec):

1. A value representing starting line, relative to the previous starting line value. If this is the first value, it is an absolute value.
2. A value representing the absolute starting column number.
3. A value representing the ending line, **relative to the starting line** value.

> 4. A value representing the ending starting column number.

Reviewer 2: There may be similar considerations for the other fields too, I just opted to go a bit deeper on this one in particular

> 4. A value representing the ending starting column number.

Author: Relative ending column may be preferable here to absolute. Now that I have the TypeScript compiler change, I'm trying to get it to compile itself to see how the performance characteristics change, and we can use iteration here to check which produces smaller values. My supposition, however, is that most often, the absolute values of ending column will generally be < the absolute values of the starting column, because that's usually how we write JS:

```js
function doSomething() { // <-- starting column
 // ...
} // <-- ending column
```

That's less true for transforming already-minified files, e.g., react-dom.production.min.js.

I'll do some experiments with the TypeScript compiler and report back before we call this "completed"

Given the above list, we should then encode the above values:

```json
{
    "sources": ["example.js"],
    "scopeNames": ["Global code", "foo"],
    "scopes": [
        "[0,0,0,33,0],[0,34,2,0,1],[2,1,1,6,0]"
    ]
}
```

Or, when encoded with VLQ, as:

```json
{
    "sources": ["example.js"],
    "scopeNames": ["Global code", "foo"],
    "scopes": [
        "AAAiCA,AkCEAC,ECCMA"
    ]
}
```

### Specific naming recommendations

The following recommendations are **recommendations only** and not part of the "standard". They are, rather, a best attempt to identify how common examples can produce high-fidelity diagnostics experiences.

> * **Class names included**: Class member functions should indicate that they are members of classes. Although a common JavaScript convention has been to use `#` as a shorthand for `Class.prototype.member`, the inclusion of `#` as valid syntax for `private` field shorthand suggests that this should not be used. This spec recommends that the expansion `ClassName.functionName` should be used for instance functions, and `static ClassName.functionName` for static functions.

Reviewer 2: Contrasting this with pasta-sourcemaps (from [here](https://github.com/bloomberg/pasta-sourcemaps/blob/master/tests/fixtures/parser/class.ts.function_descs.json)):

method: `ClassName.prototype.methodName`
getter: `ClassName.prototype.get getterName` (similar for setter)
static method: `ClassName.staticMethodName`
static getter: `ClassName.get staticGetterName` (similar for setter)

Author: I think in the first pass of the TS compiler, I rendered these as:

method: `ClassName.methodName`
getter: `ClassName.get getterName`
static method: `static ClassName.staticMethodName`
static getter: Not sure that I handled this specific one, but it would have been rendered `static ClassName.get getterName`

I'm open to changing though.

* **Anonymous callback name inference**: Callback functions are often anonymous, but the name of the function being invoked is typically interesting. In the following example, an interesting function name might be something along the lines of `anonymous function passed to map`:

```ts
    return myArray.map((item, index) => {
        /// ...
    });
```

 * **Computed names expressions**: Computed names also present an interesting problem of data that are not available at build time. As best as possible, the original source text should be simply preserved:

 ```ts
class Example {
    static [Symbol.species]() { return Object; }
    ['to' + 'String']() {
        return 'Hello, world.';
    }
}
 ```

> * In the above example, we would want to see `static Example.[Symbol.species]` and `Example.['to' + 'String']` in the `scopeNames` field of the source map.

Reviewer 2: Contrasting with pasta-sourcemaps (from [here](https://github.com/bloomberg/pasta-sourcemaps/blob/9d2022adbd87c76af1d10a352963ff1765c52def/src/parser.ts#L446-L456)):

Looks like it yields
- `<computed: identifier>`
- `<computed: stringLiteral>`
- `<computed: numericLiteral>`
- `<computed>` (else this)
